src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,1041 @@
+  _sampler = new JfrCPUTimeThreadSampler(rate, autoadapt, trace_stack_size, JfrOptionSet::stackdepth());
+  _sampler->start_thread();
+  _sampler->enroll();
+}
+
+void JfrCPUTimeThreadSampling::update_run_state(double rate, bool autoadapt) {
+  if (rate != 0) {
+    if (_sampler == nullptr) {
+      create_sampler(rate, autoadapt);
+    } else {
+      _sampler->set_rate(rate, autoadapt);
+      _sampler->enroll();
+    }
+    return;
+  }
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate /* 0 */, autoadapt);
+    _sampler->disenroll();
+  }
+}
+
+void JfrCPUTimeThreadSampling::set_rate(double rate, bool autoadapt) {
+  assert(rate >= 0, "invariant");
+  if (_instance == nullptr) {
+    return;
+  }
+  instance().set_rate_value(rate, autoadapt);
+}
+
+void JfrCPUTimeThreadSampling::set_rate_value(double rate, bool autoadapt) {
+  if (_sampler != nullptr) {
+    _sampler->set_rate(rate, autoadapt);
+  }
+  update_run_state(rate, autoadapt);
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_create(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_safepoint(JavaThread *thread) {
+  if (_sampler != nullptr) {
+    _sampler->on_safepoint(thread);
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  _instance->handle_timer_signal(info, context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(siginfo_t* info, void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers);
+  _sampler->handle_timer_signal(info, context);
+  Atomic::dec(&_sampler->_active_signal_handlers);
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(siginfo_t* info, void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrThreadLocal* jtl = jt->jfr_thread_local();
+  if (jtl == nullptr) {
+    return;
+  }
+  if (jt->thread_state() != _thread_in_native) {
+    jtl->set_wants_cpu_time_out_of_safepoint_recording(false);
+  }
+  if (!jtl->acquire_cpu_time_jfr_enqueue_lock()) {
+    jtl->increment_cpu_time_lost_samples();
+    return;
+  }
+  NoResourceMark nrm;
+  JfrCPUTimeTrace* trace = _frame_store.get_fresh_trace();
+  if (trace != nullptr) {
+    // the sampling period might be too low for the current Linux configuration
+    // so samples might be skipped and we have to compute the actual period
+    int64_t period = get_sampling_period() * (info->si_overrun + 1);
+    trace->record_trace(jt, context, period);
+    if (jtl->cpu_time_jfr_stack().enqueue(trace)) {
```

Comment:
If this fails shouldn’t we return the trace to the `_frame_store`?

Reply:
Good catch